x/examples/test-connectivity/main.go

```diff
@@ -167,15 +184,30 @@ func main() {
 		for _, proto := range strings.Split(*protoFlag, ",") {
 			proto = strings.TrimSpace(proto)
 			var resolver dns.Resolver
+			var connectionAddress string
 			switch proto {
 			case "tcp":
-				streamDialer, err := config.NewStreamDialer(*transportFlag)
+				baseDialer := transport.FuncStreamDialer(func(ctx context.Context, addr string) (transport.StreamConn, error) {
+					conn, err := (&transport.TCPDialer{}).DialStream(ctx, addr)
+					if conn != nil {
+						connectionAddress = conn.RemoteAddr().String()
+					}
+					return conn, err
+				})
+				streamDialer, err := config.WrapStreamDialer(baseDialer, *transportFlag)
```

Comment:
I tried to intercept the streamDialer instead of the baseDialer, but it didn't work because the RemoteAddress is for the target in the PacketListenerDialer. Code: https://github.com/Jigsaw-Code/outline-sdk/compare/fortuna-address2?expand=1

It's not clear what the RemoteTarget should be, so I'm not sure that is a god approach. I think I prefer to keep this base dialer interception. But we could think about some sort of event listener instead, to reduce wrapping.

Reply:
@fortuna I like the idea of intercept the streamDialer. btw in the suggested implementation, if connect fails and conn is nil, connectionAddress will be left empty and we don't know what endpoint was attempted. I can address this by extracting attempted endpoint from err by assessing error type with `err.(*net.OpError)`.

Reply:
You have the dialed address in the transport specification.

Reply:
I updated the PR. We now have functions that can test dialer wrappers, since the one that takes the resolver was not enough.

Reply:
Oh, I see your point. We won't have the address for errors, so it's hard to tell if it's a bad resolution, for instance.

I think we can extend the code to do Happy Eyeballs or [Dialer.ControlContext](https://pkg.go.dev/net#Dialer). But perhaps in another PR.

Reply:
If multiple endpoints are attempted, which one do we select? The first one? The last one? All of them? Perhaps we need to collect all of them, then specify which one we picked for the transport part.

I don't think we are able to capture all of them by looking at the error. We need to do the resolution ourselves.

Reply:
It is still a bit clunky, but I think the data model is better aligned. Example:

```
% go run . -transport $KEY -proto tcp | jq
{
  "resolver": "8.8.8.8:53",
  "proto": "tcp",
  "transport": "ss://REDACTED@159.203.76.146.nip.io:80?prefix=",
  "connections": [
    {
      "address": {
        "host": "159.203.76.146",
        "port": "80"
      },
      "error": null
    }
  ],
  "selected_address": {
    "host": "159.203.76.146",
    "port": "80"
  },
  "time": "2024-04-24T02:22:09Z",
  "duration_ms": 30,
  "error": {
    "op": "receive",
    "msg": "chacha20poly1305: message authentication failed"
  }
}
```